### ISSUE TITLE
auto-sklearn is grossly overfitting on classification model

### ISSUE THREAD

**Issue by kasuteru** [automl/auto-sklearn#633]
I recently tried auto-sklearn for the first time, on a biased binary classification dataset (~7% positives). I have ~100k samples with ~40 features per sample, and let AutoML train for 4 hours on a powerful Azure VM (~128GB RAM, 32 Cores).

I used the following (pretty standard) settings:

```
automl = autosklearn.classification.AutoSklearnClassifier(
    #include_estimators=["random_forest", ],
    exclude_estimators=None,
    ml_memory_limit=50000, 
    #include_preprocessors=["no_preprocessing", ],
    exclude_preprocessors=None,
    time_left_for_this_task=10800, 
    resampling_strategy="cv",
    resampling_strategy_arguments={'folds': 5},
    per_run_time_limit=300)

automl.fit(X_train[x_cols].copy(),
           y_train.copy(),
           metric = autosklearn.metrics.accuracy)
```

The accuracy on the training dataset was 100%. Not a single misclassification. In another run (going for balanced_accuracy instead of accuracy), there was a single misclassification out of the 70k in the training dataset.

In the test dataset however, the accuracy was barely above the baseline of 93% accuracy: 93.3% vs 93.2 baseline:
```

Confusion matrix on train set: 
[[ 5706     0] 
 [    0 69161]]

 Confusion matrix on test set: 
 [[ 612   691]
  [ 595 17308]]
```

Is there anything I am doing wrong, a parameter like dropout or regularization I should be setting? Or might this be a bug in auto-sklearn? As far as I unterstand it, auto-sklearn should interally prevent overfitting without me settings any parameters?

**Comment by mfeurer** [comment]
Hi @kasuteru thanks for bringing this up. I think there are two reasons why this happens:

1. Auto-sklearn has no measure to detect and prevent overfitting. The only thing you could do is to increase the number of CV folds.
2. Given the imbalancendness of your dataset, accuracy is most likely the wrong metric to optimize for. The choice of metric depends on your dataset and what you want to do with the outcomes, but some metrics like area under curve, f1 or balanced accuracy are definitively better suited here. There is a nice [paper by Provost, Fawcett and Kohavi](https://pdfs.semanticscholar.org/7770/3a2783f64dfceb638aa9eebd9c9c501bb835.pdf) on this issue, but I think there is no dataset independent answer to this.

**Comment by kasuteru** [comment]
Thanks for your quick answer @mfeurer! I let this algorithm run three times, optimizing for accuracy, balanced_accuracy and f1 respectively. Unfortunately, the results were similar to each other. 

1. How can I understand your point 1.? I know that H2O's AutoML internally splits the data into train/validation, and only uses the validation dataset (not used for training the algorithms, this is done with crossvalidation) for choosing hyperparameters. If I understand your answer correctly, this is not done in auto-sklearn? 

2. Does this mean that if I let the algorithm run for "too long", it will at some point overfit? Would another solution to prevent overfitting be to actually limit the runtime of the algorithm? It feels like this shouldn't be the case...

3. Another thing I think of is limiting the ensemble size... If my unterstanding is correct, this should also limit the ability of the model to overfit, or am I missing something? I will try this tonight and report on the results.

**Comment by mfeurer** [comment]
1. Auto-sklearn splits the data into training and validation, too. If there is overfitting to the validation set it cannot detect this (the same will hold true for H2O). If you're not experiencing this issue with H2O this would be great to know.

2. Most likely. This is sometimes also called oversearch. This is still an open issue. I would go with more cross validation folds first before limiting the runtime. Ultimately, one would like to have nested cross-validation to avoid this issue.

3. It could, but this depends on whether the base models already overfit the data. If they do, the ensemble will have a hard time correcting for this. If they don't they're usually doing a good job. You could post logs of Auto-sklearn to allow for further answers.

**Comment by kasuteru** [comment]
Thanks for your answers! Actually, H2O shows the same overfitting if I let it run for too long. I researched a little bit and this seems to be a general issue with AutoML algorithms. Nested crossvalidation is the solution I found as well.

Since this is not an issue with auto-sklearn but with AutoML algorithms in general, I think we can close this issue.
